python/pdo/contract/response.py

```diff
@@ -90,6 +100,216 @@ class ContractResponse(object) :
+                        cls.transaction_submission_exceptions_queue.put("Did not get a transaction id after transaction submission")
+                        raise Exception("Did not get a transaction id after transaction submission,  request nunmber %d", request_number)
+                except Exception as e:
+                    logger.error("Transaction submission failed, request nunmber %d %s", request_number, str(e))
+                    cls.transaction_submission_exceptions_queue.put("Transaction submission failed")
+                    raise Exception("Transaction submission failed, request nunmber %d %s", request_number, str(e))
+            
+            return submitted_any
+
+        # -------------------------------------------------------
+
+        rep_completed_but_txn_not_submitted_updates = dict() # key is contract_id, value is dict(k:v). k = request_number from the commit _id
+        # and v is everything else needed to submit transaction
+        
+        poll_next = True
+        
+        while poll_next:
+            task = cls.pending_transactions_queue.get()
+
+            # check if the task corresponds to a termination signal (indicating an exception of any kind)
+            if task.get('exit_now'):
+                logger.info("Exiting transaction submission woker thread")
+                break
+
+            commit_id = task['commit_id']
+            contract_id = commit_id[0]
+            request_number = commit_id[2]
+            
+            if rep_completed_but_txn_not_submitted_updates.get(contract_id):
+                rep_completed_but_txn_not_submitted_updates[contract_id][request_number] =  task
+            else:
+                rep_completed_but_txn_not_submitted_updates[contract_id] = dict({request_number: task})
+            
+            # submit transactions as many as possible for the contract_id just added 
+            try:
+                submitted_any = submit_doable_transactions_for_contract(contract_id)
+            except Exception as e:
+                cls.transaction_submission_exceptions_queue("Failure in submit_doable_transactions_for_contract")
+                logger.error("Failure in submit_doable_transactions_for_contract, %s", str(e))
+                raise Exception("Failure in submit_doable_transactions_for_contract, %s", str(e))
+                                
+            # loop over all contracts_ids. For each check contract_id, submit transactions as many as possible.
+            # Continue looping until no transaction can be submitted for any conrtract_id
+            if submitted_any and len(rep_completed_but_txn_not_submitted_updates.keys()) > 1:
+                while True:
+                    loop_again = False
+                    
+                    for contract_id in list(rep_completed_but_txn_not_submitted_updates.keys()):
+                        try:
```

Comment:
why convert this to a list? keys() already creates an iterable

Reply:
thank you, I have fixed this